### ISSUE TITLE
MethodDelegation.toField appears to not use inherited methods

### ISSUE THREAD

**Issue by bokken** [raphw/byte-buddy#214]
I am trying to generate a class where the hashCode method returns the hashCode of one of the member variables.

```
.method(HASH_CODE)
.intercept(MethodDelegation.toField(ACTUAL_VAR_NAME))
```

This leads to an IllegalArgumentException:

> java.lang.IllegalArgumentException: None of [public abstract com.cerner.mmf.interceptors.test.util.SomeInterface com.cerner.mmf.interceptors.test.util.SomeInterface.clone(), public abstract void com.cerner.mmf.interceptors.test.util.SomeInterface.ioException() throws java.io.IOException, public abstract void com.cerner.mmf.interceptors.test.util.SomeInterface.method1(java.lang.String,java.lang.Object,int,short,double,float,long), public abstract java.lang.String com.cerner.mmf.interceptors.test.util.SomeInterface.method2()] allows for delegation from public native int java.lang.Object.hashCode()
> 	at net.bytebuddy.implementation.bind.MethodDelegationBinder$Processor.process(MethodDelegationBinder.java:881)

**Comment by raphw** [comment]
Are you sure that your intended interceptor can intercept `hashCode`? Does it return `int` or uses `@RuntimeType`?

This works fine for me:

```java
new ByteBuddy()  
  .subclass(Object.class)
  .defineField("foo", Bar.class)
  .method(isHashCode())
  .intercept(MethodDelegation.toField("foo"))
  .make();
```
when using:
```
class Foo {
  @RuntimeType Object intercept() { return 42; }
}
class Bar extends Foo {

}

**Comment by bokken** [comment]
I want it to delegate it to the hashCode method inherited from Object.

**Comment by raphw** [comment]
```java
        Class<?> type = new ByteBuddy()
                .subclass(Object.class)
                .defineField("foo", Object.class, Visibility.PUBLIC)
                .method(isHashCode())
                .intercept(MethodDelegation.toField("foo").filter(isHashCode()))
                .make()
                .load(Object.class.getClassLoader())
                .getLoaded();

        Object instance = type.newInstance(), delegate = new Object();
        type.getDeclaredField("foo").set(instance, delegate);

        System.out.println(instance.hashCode());
        System.out.println(delegate.hashCode());
```

**Comment by bokken** [comment]
Is the punchline that MethodDelegation toField will only delegate to public methods declared on that field's type and not inherited methods?

Is it not any inherited methods or limited to not methods on Object?

**Comment by raphw** [comment]
With `MethodDelegation.toField("foo")`, Byte Buddy locates the field's type which in the above example is `Object`. It then finds a suitable delegation method of the field's type. As we explicitly filtered to only consider `Object::hashCode` as a possible delegate (this would have been the best match anyways), this will be used as a target for the delegation. 

By *inherited methods*, do you mean you want Byte Buddy to locate methods of the `foo` instance at runtime? This is not possible. Byte Buddy generates types where methods are dispatched virtually. If you wanted different methods for different types, you need to generate a type for each such type.

**Comment by bokken** [comment]
My foo is not of type Object, it is some type which is known at runtime, In the case of the specific example I originally posted about, it is:

```
public interface SomeInterface {

    public void method1(String string, Object obj, int i, short s, double d, float f, long l);
    
    public String method2();
    
    public SomeInterface clone();

    public void ioException() throws IOException;
}
```

Any instance of SomeInterface will also be an Object. The Exception message I get, however, lists the methods explicitly on SomeInterface and states no match can be found for hashCode. 

If I add the filter to the MethodDelegation for hashCode method (as shown in your example above) I get a message stating "None of [] allows for delegation from public native int java.lang.Object.hashCode()". This seems to indicate that only methods declared on the type are being examined. Is this expected?

**Comment by raphw** [comment]
Ah, I understand this now. While interfaces are implicitly `Object` types, they are not considered explicit subclasses of it, i.e. `someInterface.getSuperClass()` will return `null` for an interface type. It is neither possible to override `hashCode`, `toString` or `equals` with an interface default method. 

Therefore, the `Object` methods are not considered here. There are two ways around it:
1. You declare the methods in the interface.
2. You implement a custom `MethodGraph.Compiler` that returns a method graph where these methods are included.

I will at some point add a convenience method to the latter.

**Comment by raphw** [comment]
I revisited this and I wonder if `MethodDelegation` is the right approach for this. Interfaces do, in theory, not define the `Object` methods and I would break a lot of things by making Byte Buddy think so. Are you trying to redirect an object's methods to another object? I think it would be better to implement this with an explicit call, i.e.:
```java
.method(HASH_CODE)
.intercept(MethodCall.invoke(Object.class.getDeclaredMethod("hashCode"))
```
I am still looking into a more convenient solution but for the time being, I think this is the best way.

**Comment by bokken** [comment]
I am trying to redirect the call to member variable.

**Comment by raphw** [comment]
In this case, you can, for now, use `MethodCall.invoke(Object.class.getDeclaredMethod("hashCode").onField("foo"))`.

I added this in version 1.5.8 (released yesterday). I hope to come up with something smarter in the future.

**Comment by raphw** [comment]
After looking into this, I decided to not add such functionality. The only time where I would like to use a delegation to the `Object` methods would be when I use a delegation to delegate every method from instance A to instance B where both instances are of the same type. For such cases, it would be best to use the suggested `MethodCall` which is much better performant and which can be chained. I added a reference to `MethodCall` in the documentation.
